gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,972 @@
+}
+
+static gboolean
+eos_reformatter_decompress (EosReformatter *reformatter, GAsyncQueue *inq, GAsyncQueue *outq)
+{
+  GConverterResult res;
+  GConverterFlags flags = G_CONVERTER_NO_FLAGS;
+  GError *error = NULL;
+  gsize input_read = 0;
+  EosAlignedBuffer *buf;
+  EosAlignedBuffer *outbuf;
+
+  if (reformatter->error != NULL)
+    {
+      /* Error is set so send EOS and abort */
+      buf = g_new0(EosAlignedBuffer, 1);
+      g_async_queue_push (outq, buf);
+      return FALSE;
+    }
+
+  buf = g_async_queue_pop (inq);
+  outbuf = eos_reformatter_get_free_buffer (reformatter, TRUE);
+
+  if (buf->len == 0)
+    flags |= G_CONVERTER_INPUT_AT_END;
+
+  outbuf->len = 0;
+
+  /* Loop over input until consumed, sending possible output immediately */
+  do
+    {
+      gsize in = 0;
+      res = g_converter_convert (reformatter->decompressor,
+                                 buf->ptr + input_read, buf->len - input_read,
+                                 outbuf->ptr, reformatter->buffer_size,
+                                 flags, &in, &outbuf->len,
+                                 &error);
+      input_read += in;
+
+      if (outbuf->len > 0)
+        {
+          g_async_queue_push (outq, outbuf);
+          /* Note that this competes with read buffers if in different threads */
+          outbuf = eos_reformatter_get_free_buffer (reformatter, TRUE);
+          if (reformatter->error != NULL)
+            {
+              res = G_CONVERTER_ERROR;
+            }
+        }
+    }
+  while (res == G_CONVERTER_CONVERTED && input_read < buf->len);
+
+  /* Return the outbuf to free queue if we didn't use it */
+  if (outbuf->len == 0)
+    {
+      g_async_queue_push (reformatter->free_queue, outbuf);
+    }
+
+  /* This is fine, we just need to cycle to get more input buffers */
+  if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_PARTIAL_INPUT))
+    {
+      res = G_CONVERTER_CONVERTED;
+    }
+
+  if (error != NULL)
+    {
+      reformatter->error = g_error_new (EOS_REFORMATTER_ERROR, 0, _("Internal error"));
+      g_error_free (error);
+    }
+
+  if (res == G_CONVERTER_FINISHED)
+    {
+      /* If conversion is finished, we send an empty buffer to indicate EOS */
+      buf->len = 0;
+      g_async_queue_push (outq, buf);
+    }
+  else
+    {
+      g_async_queue_push (reformatter->free_queue, buf);
+    }
+
+  return (res == G_CONVERTER_CONVERTED);
+}
+
+static gint
+eos_reformatter_prepare_write (EosReformatter *reformatter)
+{
+  return open(reformatter->device, O_WRONLY | O_SYNC | O_CREAT);
+}
+
+static gboolean
+eos_reformatter_add_sample (gpointer data)
+{
+  EosReformatter *reformatter = EOS_REFORMATTER(data);
+
+  gdu_estimator_add_sample (reformatter->estimator, reformatter->write_completed_bytes);
+
+  return FALSE;
+}
+
+static gboolean
+eos_reformatter_write (EosReformatter *reformatter, GAsyncQueue *inq)
+{
+  gboolean ret = TRUE;
+  EosAlignedBuffer *buf;
+  gsize wb = 0;
+
+  if (reformatter->error != NULL)
```

Comment:
Maybe use a `GCancellable` to indicate the case when all the threads can stop. As a bonus this might let us add a Cancel button (of dubious utility of course…).